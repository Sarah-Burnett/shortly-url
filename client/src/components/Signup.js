import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import { PrimaryFormBtn, Form, Label, Input } from '../Styles';
import { Errors } from './AuthErrors';
import axios from 'axios';

export const Signup =  () => {
    const [ formValues, setFormValues ] = useState({
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        password2: '',
    });
    const { firstName, lastName, email, password, password2, errors } = formValues;
    const handleChange = (event) => {
        event.persist()
        setFormValues(values => ({...values, [event.target.id]: event.target.value}))   
    };
    const signup = (event) => {
        event.preventDefault()
        axios
          .post('/api/users/signup', {firstName, lastName, email, password})
          .then(() => <Redirect to="/login" />)
          .catch(err => err)
    };
    return (
        <Form onSubmit={signup}>
            <h3>Sign Up</h3>
            { formValues.error && <Errors errors={errors}/> }
            <Label HTMLfor="firstName">First Name <Input id="firstName" name="firstName" onChange={handleChange} value={firstName} required></Input></Label>
            <Label HTMLfor="lastName">Last Name <Input id="lastName" name="lastName" onChange={handleChange} value={lastName} required></Input></Label>
            <Label HTMLfor="email">Email <Input type="email" id="email" name="email" onChange={handleChange} value={email} required></Input></Label>
            <Label HTMLfor="password">Password <Input type="password" id="password" name="password" onChange={handleChange} value={password} required></Input></Label>
            <Label HTMLfor="password2">Confirm your password <Input type="password" id="password2" name="password2" onChange={handleChange} value={password2} required></Input></Label>
            <PrimaryFormBtn type="submit">Sign up</PrimaryFormBtn>
        </Form>
    );
}