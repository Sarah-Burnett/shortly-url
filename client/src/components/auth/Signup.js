import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import { Errors } from './AuthErrors';
import axios from 'axios';
import { isSignupValid } from '../../utilities/validateAuth';
import { AuthForm, Label, Input, FormBtn, PrimaryFormBtn } from '../../Styles';

export const Signup =  () => {
    const [ formValues, setFormValues ] = useState({
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        password2: '',
    });
    const [ errors, setErrors ] = useState([]);
    const { firstName, lastName, email, password, password2 } = formValues;
    const handleChange = (event) => {
        event.persist()
        setFormValues(values => ({...values, [event.target.id]: event.target.value}))   
    };
    const signup = (event) => {
        event.preventDefault();
        const errorMsgs = isSignupValid(formValues);
        console.log(errorMsgs.length);
        if (errorMsgs.length !== 0) return setErrors(errorMsgs);
        axios
          .post('/api/users/signup', {firstName, lastName, email, password})
          .then(() => <Redirect to="/login" />)
          .catch(err => setErrors([err.response.data.error || "Sign up failed. Please try again"]))
    };
    return (
        <AuthForm onSubmit={signup} noValidate>
            <h3>Sign Up</h3>
            { errors.length > 0 && <Errors errors={errors}/> }
            <Label HTMLfor="firstName">First Name <Input id="firstName" name="firstName" onChange={handleChange} value={firstName} required></Input></Label>
            <Label HTMLfor="lastName">Last Name <Input id="lastName" name="lastName" onChange={handleChange} value={lastName} required></Input></Label>
            <Label HTMLfor="email">Email <Input type="email" id="email" name="email" onChange={handleChange} value={email} required></Input></Label>
            <Label HTMLfor="password">Password <Input type="password" id="password" name="password" onChange={handleChange} value={password} pattern={password2} required></Input></Label>
            <Label HTMLfor="password2">Confirm your password <Input type="password" id="password2" name="password2" onChange={handleChange} value={password2} pattern={password} required></Input></Label>
            <PrimaryFormBtn type="submit">Sign up</PrimaryFormBtn>
        </AuthForm>
    );
}