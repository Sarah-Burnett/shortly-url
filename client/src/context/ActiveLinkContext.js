import React, { useState, createContext } from 'react';

export const ActiveLinkContext = createContext();

export const ActiveLinkProvider = (props) => {
    const [ activeLink, setActiveLink] = useState({
        activeLink: [],
        process: false
    });

    return (
        <ActiveLinkContext.Provider value={[activeLink, setActiveLink]}>
            {props.children}
        </ActiveLinkContext.Provider>
    )
};