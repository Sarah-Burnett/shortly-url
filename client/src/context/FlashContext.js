import React, { useState, createContext } from 'react';

export const FlashContext = createContext();

export const FlashProvider = (props) => {
    const [ flash, setFlash ] = useState({
        flashMsg: "",
        isFlashVisible: false,
        isAutoHideFlash: true,
    });
    return (
        <FlashContext.Provider value={[flash, setFlash]}>
            {props.children}
        </FlashContext.Provider>
    )
};