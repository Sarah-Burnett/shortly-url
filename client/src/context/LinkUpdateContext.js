import React, { useState, createContext } from 'react';

export const LinkUpdateContext = createContext();

export const LinkUpdateProvider = (props) => {
    const [ linkUpdate, setLinkUpdate] = useState();
    return (
        <LinkUpdateContext.Provider value={[linkUpdate, setLinkUpdate]}>
            {props.children}
        </LinkUpdateContext.Provider>
    )
};