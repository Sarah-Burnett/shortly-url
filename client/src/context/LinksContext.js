import React, { useState, createContext } from 'react';

export const LinksContext = createContext();

export const LinksProvider = (props) => {
    const [ links, setLinks] = useState([]);

    return (
        <LinksContext.Provider value={[links, setLinks]}>
            {props.children}
        </LinksContext.Provider>
    )
};