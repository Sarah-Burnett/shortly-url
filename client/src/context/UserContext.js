import React, { useState, createContext } from 'react';

export const UserContext = createContext();

export const UserProvider = (props) => {
    const [ user, setUser] = useState({
        isAuth: false,
        firstName: undefined,
        id: undefined,
        token: undefined,
    });
    
    return (
        <UserContext.Provider value={[user, setUser]}>
            {props.children}
        </UserContext.Provider>
    )
};