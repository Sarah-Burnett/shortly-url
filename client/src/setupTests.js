import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
