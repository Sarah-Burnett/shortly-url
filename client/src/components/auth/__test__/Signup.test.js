import React from 'react';
import { rendersWithoutCrashing, matchesSnapshot } from '../../../setupTests';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import Signup from '../Signup';
import { UserProvider } from '../../../context/UserContext';
import { LinksProvider } from '../../../context/LinksContext';
import { FlashProvider } from '../../../context/FlashContext';

const SignupWrapper = (formValues) => {
    return (
        <UserProvider>
            <LinksProvider>
                <FlashProvider>
                    <Signup formValues={formValues} setFormValues={() => {}} signupUrl={'/signup'}/>
                </FlashProvider>
            </LinksProvider>
        </UserProvider>
    )
}

const ValidSignupWrapper = () => SignupWrapper({
    firstName: "Harriet",
    lastName: "Tripp",
    email: "harriet-tripp@outlook.com", 
    password: "123456", 
    password2: "123456", 
})


const InvalidSignupWrapper = () => SignupWrapper({
    firstName: "Harriet",
    lastName: "",
    email: "", 
    password: "1234", 
    password2: "", 
})

const server = setupServer(
    rest.post('/signup', (req, res, ctx) => {
        return res(ctx.json({
            "_id":"5f3a49088a296b42c0089d10",
            "firstName":"Harriet",
            "lastName":"Tripp",
            "email":"harriet.tripp@outlook.com",
            "password":"$2a$10$fa0d9DBBN96GR.i9HU/U7uSe08s4Zl6Ik8t9GC/mw9j7VxaDPJoNi",
            "date":"2020-08-17T09:08:24.192Z",
            "__v":0
        }))
    })
)
  
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterEach(cleanup);
afterAll(() => server.close())
 
//TODO:
// test 404 server error - err.response.data.error undefined

it('renders signup without crashing', () => rendersWithoutCrashing(ValidSignupWrapper));

it('signup matches snapshot', () => matchesSnapshot(ValidSignupWrapper));

it('valid user is signed up', async () => {
    render(<ValidSignupWrapper/>);
    await fireEvent.click(screen.getByText('Sign up'));
    expect(screen.queryByTestId('errors')).toBe(null);
});

it('invalid user shows error', async () => {
    render(<InvalidSignupWrapper/>);
    fireEvent.click(screen.getByText('Sign up'));
    await waitForElement(() => screen.getByTestId('errors'))
    expect(screen.getByTestId('errors')).toHaveTextContent('Please enter your last namePlease enter your emailPlease confirm your passwordPassword must be at least 6 charactersEntered passwords are not the same');
})

it('handles server error', async () => {
    server.use(
        rest.post('/signup', (req, res, ctx) => {
        return res(ctx.status(500))
        })
    )
    render(<ValidSignupWrapper/>);
    fireEvent.click(screen.getByText('Sign up'));
    await waitForElement(() => screen.getByTestId('errors'))
    expect(screen.getByTestId('errors')).toHaveTextContent('Sign up failed. Please try again');
});

