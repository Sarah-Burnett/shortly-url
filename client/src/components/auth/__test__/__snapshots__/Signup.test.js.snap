// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup matches snapshot 1`] = `
<form
  className="sc-AxhUy dLXkxw"
  data-testid="signup"
  noValidate={true}
  onSubmit={[Function]}
>
  <h3>
    Sign Up
  </h3>
  <label
    className="sc-Axmtr fYuxyA"
  >
    First Name
     
    <input
      className="sc-AxmLO erTjTr"
      id="firstName"
      name="firstName"
      onChange={[Function]}
      required={true}
      value="Harriet"
    />
  </label>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Last Name
     
    <input
      className="sc-AxmLO erTjTr"
      id="lastName"
      name="lastName"
      onChange={[Function]}
      required={true}
      value="Tripp"
    />
  </label>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Email
     
    <input
      className="sc-AxmLO erTjTr"
      id="email"
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value="harriet-tripp@outlook.com"
    />
  </label>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Password
     
    <input
      className="sc-AxmLO erTjTr"
      id="password"
      name="password"
      onChange={[Function]}
      pattern="123456"
      required={true}
      type="password"
      value="123456"
    />
  </label>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Confirm your password
     
    <input
      className="sc-AxmLO erTjTr"
      id="password2"
      name="password2"
      onChange={[Function]}
      pattern="123456"
      required={true}
      type="password"
      value="123456"
    />
  </label>
  <button
    className="sc-AxirZ sc-AxheI cOblFm"
    type="submit"
  >
    Sign up
  </button>
</form>
`;
